Cython/Compiler/ModuleNode.py

```diff
@@ -1795,9 +1795,15 @@ def generate_dealloc_function(self, scope, code):
                 # cimported base type pointer directly interacts badly with
                 # the module cleanup, which may already have cleared it.
                 # In that case, fall back to traversing the type hierarchy.
-                code.putln("if (likely(%s)) __Pyx_PyType_GetSlot(%s, tp_dealloc, destructor)(o); "
-                           "else __Pyx_call_next_tp_dealloc(o, %s);" % (
-                               base_cname, base_cname, slot_func_cname))
+                # If we're using the module state then always go through the
+                # type hierarchy, because our access to the module state may
+                # have been lost (at least for the limited API version of
+                # using module state).
+                code.putln("#if !CYTHON_USE_MODULE_STATE")
```

Comment:
Definitely module state.

I think the crash was in `PyState_FindModule` in destructors being called as Python shut down.

It's a much more general problem of course, since destructors can call arbitrary Python code which might need to access module state itself. But this covers a very common case in the Cython-generated part of the code.